planner/core/preprocess_test.go

```diff
@@ -220,28 +258,23 @@ func (s *testValidatorSuite) TestValidator(c *C) {
 		{"CREATE TABLE origin (a int key auto_increment, b int);", false, nil},
 	}
 
-	store, dom, err := newStoreWithBootstrap()
+	_, err := s.se.Execute(context.Background(), "use test")
 	c.Assert(err, IsNil)
-	defer func() {
-		dom.Close()
-		store.Close()
-	}()
-	se, err := session.CreateSession4Test(store)
-	c.Assert(err, IsNil)
-	_, err = se.Execute(context.Background(), "use test")
-	c.Assert(err, IsNil)
-	ctx := se.(sessionctx.Context)
-	is := infoschema.MockInfoSchema([]*model.TableInfo{core.MockSignedTable()})
+
 	for _, tt := range tests {
-		stmts, err1 := session.Parse(ctx, tt.sql)
-		c.Assert(err1, IsNil)
-		c.Assert(stmts, HasLen, 1)
-		stmt := stmts[0]
-		var opts []core.PreprocessOpt
-		if tt.inPrepare {
-			opts = append(opts, core.InPrepare)
-		}
-		err = core.Preprocess(ctx, stmt, is, opts...)
-		c.Assert(terror.ErrorEqual(err, tt.err), IsTrue, Commentf("sql: %s, err:%v", tt.sql, err))
+		s.runSQL(c, tt.sql, tt.inPrepare, tt.err)
 	}
 }
+
+func (s *testValidatorSuite) TestForeignKey(c *C) {
+	defer testleak.AfterTest(c)()
+
+	s.runSQL(c, "ALTER TABLE test.t ADD CONSTRAINT fk FOREIGN KEY (c2) REFERENCES t (c1)", false, nil)
+
+	_, err := s.se.Execute(context.Background(), "use test")
+	c.Assert(err, IsNil)
+
+	s.runSQL(c, "ALTER TABLE test.t ADD CONSTRAINT fk FOREIGN KEY (c3) REFERENCES t (c1)", false, nil)
+
+	s.runSQL(c, "ALTER TABLE test.t ADD CONSTRAINT fk FOREIGN KEY t3(c3) REFERENCES t (c1)", false, nil)
```

Comment:
I mean, `REFERENCES test2.t(c1)`, where `test2` is another schema.

Reply:
Oh, that is a mistake.

Reply:
Are you sure `test2.t` exists?

Reply:
It seems that the underlying store is read only. Since the code here is not actually executed, but only parsed/resolved. I think it is ok to just use `test.t`.

Reply:
So you can't judge from the test case that whether `test` is read from `ALTER TABLE test.t` or `REFERENCES test.t(c1)`.